docs/specs/oidc-native-sso.md

```diff
@@ -0,0 +1,153 @@
+    + [`grant_type=authorization_code`](#grant_typeauthorization_code)
+    + [`grant_type=urn:authgear:params:oauth:grant-type:biometric-request`](#grant_typeurnauthgearparamsoauthgrant-typebiometric-request)
+    + [`grant_type=refresh_token`](#grant_typerefresh_token)
+    + [`grant_type=urn:authgear:params:oauth:grant-type:app2app`](#grant_typeurnauthgearparamsoauthgrant-typeapp2app)
+    + [`grant_type=urn:ietf:params:oauth:grant-type:token-exchange`](#grant_typeurnietfparamsoauthgrant-typetoken-exchange)
+    + [When issuing ID tokens](#when-issuing-id-tokens)
+  * [Changes on Admin API](#changes-on-admin-api)
+  * [Changes on SDK](#changes-on-sdk)
+    + [Recipe: Two applications written by the same vendor](#recipe-two-applications-written-by-the-same-vendor)
+    + [Recipe: A application opening webapps with custom webview](#recipe-a-application-opening-webapps-with-custom-webview)
+  * [Caveats](#caveats)
+
+# OIDC Native SSO
+
+This document specifies the implementation of [OIDC Native SSO](https://openid.net/specs/openid-connect-native-sso-1_0.html).
+
+## Key points to note
+
+- A OfflineGrant used to has only one set of client-specific information, like `client_id` and `refresh_token`. Now, a OfflineGrant can have multiple refresh tokens.
+- Since there is only one OfflineGrant, the apps sharing user authentication with Native SSO does not share refresh token. But the underlying session is shared. Thus signing out in one app will sign out all apps. This is by design. See [App2app](./app2app.md) if you want the apps have independent sessions.
+- Native SSO is done through the Token Endpoint, thus it requires no user interaction.
+- Existing OfflineGrant cannot perform Native SSO. It is because `device_sso` is not in the `scope`. Sign in again to obtain a Native SSO OfflineGrant to perform Native SSO.
+
+## Changes on Client
+
+- Add `x_device_sso_enabled: boolean`.
+- `scope=device_sso` is allowed if `x_device_sso_enabled=true`.
+
+> Do we need to add `x_device_sso_key: string` to designate which group of clients can perform Native SSO?
+> Only clients with `x_device_sso_enabled=true` AND the same value of `x_device_sso_key` can perform Native SSO with each other.
+> This seems very advanced to me.
+
+## Changes on OfflineGrant
+
+- The following fields become client-specific
+  - `ClientID`
+  - `AuthorizationID`
+  - `IdentityID`
+  - `Scopes`
+  - `TokenHash`
+- Add a new field `DeviceSSORefreshTokens`. It has the above client-specific fields.
+  - `ClientID`
+  - `AuthorizationID`
+  - `IdentityID`
+  - `Scopes`
+  - `TokenHash`
+- Add a new field `DeviceSecretHash`. It is the hex of SHA256 of `device_secret`.
+- If `DeviceSSORefreshTokens` is non-empty, then the offline grant is a Native SSO offline grant.
+- Otherwise, the offline grant is an ordinary offline grant.
+
+## Changes on Authorization Endpoint
+
+- Allow `scope=device_sso` if the client has `x_device_sso_enabled=true`.
+
+## Changes on Token Endpoint
```

Comment:
It seems that we need to support `device_sso` on anonymous user if we plan to use it for settings page in future

Reply:
I think anonymous user cannot open settings page.